erts/emulator/beam/erl_bits_f16.h

```diff
@@ -0,0 +1,202 @@
+ * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies
+ * or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * The following changes were done to adapt this to Erlang/OTP conventions:
+ *
+ * 1. uint16_t and uint32_t have been rewritten to Uint16 and Uint32
+ * 2. Mixed declarations were moved to the top to avoid warnings
+ * 3. inline was rewritten as ERTS_INLINE
+ */
+
+static ERTS_INLINE float fp32_from_bits(Uint32 w) {
+    union {
+	Uint32 as_bits;
+	float as_value;
+    } fp32 = { w };
+    return fp32.as_value;
+}
+
+static ERTS_INLINE Uint32 fp32_to_bits(float f) {
+    union {
+	float as_value;
+	Uint32 as_bits;
+    } fp32 = { f };
+    return fp32.as_bits;
+}
+
+/*
+ * Convert a 32-bit floating-point number in IEEE single-precision format to a 16-bit floating-point number in
+ * IEEE half-precision format, in bit representation.
+ *
+ * @note The implementation relies on IEEE-like (no assumption about rounding mode and no operations on denormals)
+ * floating-point operations and bitcasts between integer and floating-point variables.
+ */
+static ERTS_INLINE Uint16 fp16_ieee_from_fp32_value(float f) {
+#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L) || defined(__GNUC__) && !defined(__STRICT_ANSI__)
+    const float scale_to_inf = 0x1.0p+112f;
+    const float scale_to_zero = 0x1.0p-110f;
+#else
+    const float scale_to_inf = fp32_from_bits(UINT32_C(0x77800000));
+    const float scale_to_zero = fp32_from_bits(UINT32_C(0x08800000));
+#endif
+    float base = (fabsf(f) * scale_to_inf) * scale_to_zero;
+
+    const Uint32 w = fp32_to_bits(f);
+    const Uint32 shl1_w = w + w;
+    const Uint32 sign = w & UINT32_C(0x80000000);
+
+    Uint32 bits, exp_bits, mantissa_bits, nonsign, bias;
+
+    bias = shl1_w & UINT32_C(0xFF000000);
+
+    if (bias < UINT32_C(0x71000000)) {
+	bias = UINT32_C(0x71000000);
+    }
+
+    base = fp32_from_bits((bias >> 1) + UINT32_C(0x07800000)) + base;
+    bits = fp32_to_bits(base);
+    exp_bits = (bits >> 13) & UINT32_C(0x00007C00);
+    mantissa_bits = bits & UINT32_C(0x00000FFF);
+    nonsign = exp_bits + mantissa_bits;
+    return (sign >> 16) | (shl1_w > UINT32_C(0xFF000000) ? UINT16_C(0x7E00) : nonsign);
```

Comment:
```
[2021-02-03T10:53:10.718Z] erl_bits.o : error LNK2019: unresolved external symbol UINT32_C referenced in function erts_bs_get_float_2
[2021-02-03T10:53:10.718Z] erl_bits.o : error LNK2019: unresolved external symbol UINT16_C referenced in function fp16_ieee_from_fp32_value
```

Reply:
Thanks @jhogberg! Which platform is it coming from? It seems those macros are part of the C99 standard, so I am wondering why they are not available. Although uint16_t is part of the C99 standard too.

Reply:
Follow up: do we support any 16bit platforms? My understanding is that this will matter only for those platforms, otherwise adding a U suffix would be enough.

Reply:
A relatively old version of MSVC that doesn't fully support C99 yet. We don't support any 16-bit platforms so it should be safe to use a `U` prefix.

Reply:
Updated! I also re-ran the added tests and they continue to pass.

Reply:
Thank you, it builds fine now :)